ush/python/pygw/src/pygw/schema.py

```diff
@@ -0,0 +1,891 @@
+
+            return _create_or_use_ref(return_schema)
+
+        return _json_schema(self, True)
+
+
+class Optional(Schema):
+    """Marker for an optional part of the validation Schema."""
+
+    _MARKER = object()
+
+    def __init__(self, *args, **kwargs):
+        default = kwargs.pop("default", self._MARKER)
+        super(Optional, self).__init__(*args, **kwargs)
+        if default is not self._MARKER:
+            # See if I can come up with a static key to use for myself:
+            if _priority(self._schema) != COMPARABLE:
+                raise TypeError(
+                    "Optional keys with defaults must have simple, "
+                    "predictable values, like literal strings or ints. "
+                    '"%r" is too complex.' % (self._schema,)
+                )
+            self.default = default
+            self.key = str(self._schema)
+
+    def __hash__(self):
+        return hash(self._schema)
+
+    def __eq__(self, other):
+        return (
+            self.__class__ is other.__class__
+            and getattr(self, "default", self._MARKER) == getattr(other, "default", self._MARKER)
+            and self._schema == other._schema
+        )
+
+    def reset(self):
+        if hasattr(self._schema, "reset"):
+            self._schema.reset()
+
+
+class Hook(Schema):
+    def __init__(self, *args, **kwargs):
+        self.handler = kwargs.pop("handler", lambda *args: None)
+        super(Hook, self).__init__(*args, **kwargs)
+        self.key = self._schema
+
+
+class Forbidden(Hook):
+    def __init__(self, *args, **kwargs):
+        kwargs["handler"] = self._default_function
+        super(Forbidden, self).__init__(*args, **kwargs)
+
+    @staticmethod
+    def _default_function(nkey, data, error):
+        raise SchemaForbiddenKeyError(
+            "Forbidden key encountered: %r in %r" % (nkey, data), error)
+
+
+class Literal(object):
+    def __init__(self, value, description=None):
+        self._schema = value
+        self._description = description
+
+    def __str__(self):
+        return self._schema
+
+    def __repr__(self):
+        return 'Literal("' + self.schema + '", description="' + (self.description or "") + '")'
+
+    @property
+    def description(self):
+        return self._description
+
+    @property
+    def schema(self):
+        return self._schema
+
+
+class Const(Schema):
+    def validate(self, data, **kwargs):
+        super(Const, self).validate(data, **kwargs)
+        return data
+
+
+def _callable_str(callable_):
+    if hasattr(callable_, "__name__"):
+        return callable_.__name__
+    return str(callable_)
+
+
+def _plural_s(sized):
+    return "s" if len(sized) > 1 else ""
+
+# ----
+
+# The following functions are specific to global-workflow. The Schema
+# class module was obtained from:
+
+# https://github.com/keleshev/schema/blob/master/schema.py
+
+
+# TODO: Is there a better way to do this?
+logger = Logger("pygw.schema", colored_log=True)
```

Comment:
I don't think we need to log inside of `schema`.
```suggestion
```

Reply:
Done.